smithy-aws-cloudformation-traits/src/main/java/software/amazon/smithy/aws/cloudformation/traits/ResourceIndex.java

```diff
@@ -0,0 +1,479 @@
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.IdentifierBindingIndex;
+import software.amazon.smithy.model.knowledge.KnowledgeIndex;
+import software.amazon.smithy.model.knowledge.OperationIndex;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.ResourceShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.shapes.ShapeVisitor;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.shapes.ToShapeId;
+import software.amazon.smithy.model.traits.Trait;
+import software.amazon.smithy.utils.ListUtils;
+import software.amazon.smithy.utils.MapUtils;
+import software.amazon.smithy.utils.SetUtils;
+
+/**
+ * Index of resources to their CloudFormation identifiers
+ * and properties.
+ *
+ * <p>This index performs no validation that the identifiers
+ * and reference valid shapes.
+ */
+public final class ResourceIndex implements KnowledgeIndex {
+
+    static final Set<Mutability> FULLY_MUTABLE = SetUtils.of(
+            Mutability.CREATE, Mutability.READ, Mutability.WRITE);
+
+    private final Model model;
+    private final Map<ShapeId, Map<String, ResourcePropertyDefinition>> resourcePropertyMutabilities = new HashMap<>();
+    private final Map<ShapeId, Set<ShapeId>> resourceExcludedProperties = new HashMap<>();
+    private final Map<ShapeId, Set<String>> resourcePrimaryIdentifiers = new HashMap<>();
+    private final Map<ShapeId, List<Set<String>>> resourceAdditionalIdentifiers = new HashMap<>();
+
+    /**
+     * CloudFormation-specific property mutability options.
+     */
+    public enum Mutability {
+        CREATE,
+        READ,
+        WRITE
+    }
+
+    public ResourceIndex(Model model) {
+        this.model = model;
+
+        OperationIndex operationIndex = OperationIndex.of(model);
+        model.shapes(ResourceShape.class)
+                .flatMap(shape -> Trait.flatMapStream(shape, ResourceTrait.class))
+                .forEach(pair -> {
+                    ResourceShape resource = pair.getLeft();
+                    ShapeId resourceId = resource.getId();
+
+                    // Start with the explicit resource identifiers.
+                    resourcePrimaryIdentifiers.put(resourceId, SetUtils.copyOf(resource.getIdentifiers().keySet()));
+                    setIdentifierMutabilities(resource);
+
+                    // Use the read lifecycle's input to collect the additional identifiers
+                    // and its output to collect readable properties.
+                    resource.getRead().ifPresent(operationId -> {
```

Comment:
Not sure if it's possible to coherently refactor this in the way I'm thinking, but this code seems similar to the code used to validate properties too. I wonder if there's a way to create more of an abstraction for resolving properties and mutability using a more OO approach that separately encapsulates resolving properties (can be used in the validation stuff and probably feed into mutability resolution), mutability (replacing lots of stuff in this index), targets, etc.

Reply:
I took a smaller approach to refactoring this that handles removing the similar code and handles the potential edge cases safely.